File: schemas/project-management/convex/custom_fields.ts
// customFields: user-defined fields extending the task schema per project.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const customFields = defineTable({
  projectId: v.id("projects"),
  name: v.string(),
  fieldType: v.union(
    v.literal("text"),
    v.literal("number"),
    v.literal("date"),
    v.literal("select"),
    v.literal("multi_select"),
    v.literal("checkbox"),
    v.literal("url")
  ),
  description: v.optional(v.string()),
  isRequired: v.boolean(),
  position: v.number(),
  updatedAt: v.number(),
})
  .index("by_project_id_position", ["projectId", "position"]);

File: schemas/project-management/spacetimedb/custom_fields.rs
// custom_fields: User-defined field definitions for extending task data.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum CustomFieldType {
    Text, // type: String
    Number,
    Date,
    Select,
    MultiSelect,
    Checkbox,
    Url,
}

#[spacetimedb::table(name = custom_fields, public)]
pub struct CustomField {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub project_id: String, // UUID — FK → projects.id (cascade delete)
    pub name: String,
    pub field_type: CustomFieldType,
    pub description: Option<String>,
    pub is_required: bool,
    pub position: i32,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
    // Composite index: (project_id, position)
}
